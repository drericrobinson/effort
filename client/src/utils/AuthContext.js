import React from 'react';

const AuthContext = React.createContext();

class AuthProvider extends React.Component {
  state = {
    isAuth: localStorage.getItem('token') ? true : false,
    token: localStorage.getItem('token') ? localStorage.getItem('token') : null,
    user: localStorage.getItem('user')
      ? JSON.parse(localStorage.getItem('user'))
      : null,
    loginError: null
  };

  constructor() {
    super();
    this.login = this.login.bind(this);
    this.logout = this.logout.bind(this);
    this.isCoach = this.isCoach.bind(this);
  }

  isCoach() {
    return this.state.user.coachOfGroups.length > 0;
  }

  login(values) {
    const { email, password } = values;
    fetch('/signin', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ email, password })
    })
      .then(response =>
        response.json().then(data => ({
          status: response.status,
          token: data.token,
          user: data.user,
          message: data.message
        }))
      )
      .then(response => {
        if (response.status === 200) {
          this.setState({
            token: response.token,
            user: response.user,
            isAuth: true,
            loginError: null
          });
          localStorage.setItem('token', response.token);
          localStorage.setItem('user', JSON.stringify(response.user));
        } else {
          this.setState({ loginError: response.message });
        }
      })
      .catch(error => {
        this.setState({ loginError: error.message });
      });
  }
  logout() {
    this.setState({ isAuth: false, user: null, token: null });
    localStorage.removeItem('token');
    localStorage.removeItem('user');
  }

  render() {
    return (
      <AuthContext.Provider
        value={{
          isAuth: this.state.isAuth,
          login: this.login,
          logout: this.logout,
          loginError: this.state.loginError,
          user: this.state.user,
          isCoach: this.isCoach
        }}
      >
        {this.props.children}
      </AuthContext.Provider>
    );
  }
}

const AuthConsumer = AuthContext.Consumer;

export { AuthProvider, AuthConsumer, AuthContext };
