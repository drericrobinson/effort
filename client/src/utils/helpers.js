import { startOfWeek, endOfWeek, format } from 'date-fns';

export const weekStart = date => {
  return startOfWeek(date, { weekStartsOn: 1 });
};

export const weekEnd = date => {
  return endOfWeek(date, { weekStartsOn: 1 });
};

export const dateInWords = date => {
  return format(date, 'MMM D');
};

export const weekRangeInWords = date => {
  return (
    dateInWords(weekStart(date)) +
    String.fromCharCode(8212) +
    dateInWords(weekEnd(date))
  );
};
