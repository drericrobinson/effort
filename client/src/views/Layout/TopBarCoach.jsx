import React from 'react';
import { Link } from '@reach/router';
import { UikTopBarLinkContainer, UikTopBarLink } from '../../@uik';

export const TopBarCoach = () => {
  return (
    <UikTopBarLinkContainer>
      <UikTopBarLink Component={Link} to='/reports'>
        Reports
      </UikTopBarLink>
      <UikTopBarLink Component={Link} to='/group'>
        Group
      </UikTopBarLink>
    </UikTopBarLinkContainer>
  );
};
