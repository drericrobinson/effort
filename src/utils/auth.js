import config from '../config';
import { User } from '../resources/user/user.model';
import jwt from 'jsonwebtoken';

export const newToken = user => {
  return jwt.sign({ id: user.id }, config.secrets.jwt, {
    expiresIn: config.secrets.jwtExp
  });
};

export const verifyToken = token =>
  new Promise((resolve, reject) => {
    jwt.verify(token, config.secrets.jwt, (err, payload) => {
      if (err) return reject(err);
      resolve(payload);
    });
  });

export const signup = async (req, res) => {
  const { email, password, firstName, lastName } = req.body;

  if (!email || !password || !firstName || !lastName) {
    return res.status(400).send({
      message: 'Email, password, first name, and last name required.'
    });
  }

  try {
    const user = await User.create({ email, password, firstName, lastName });
    const token = newToken(user);

    return res.status(201).send({ token });
  } catch (e) {
    return res.status(500).send({ error: e });
  }
};

export const signin = async (req, res) => {
  const { email, password } = req.body;

  if (!email || !password) {
    return res.status(400).send({ message: 'Email and password required.' });
  }

  try {
    const user = await User.findOne({ email }).exec();
    const match = await user.checkPassword(password);

    if (!match) {
      return res.status(401).send({ message: 'Invalid password.' });
    }

    const token = newToken(user);
    return res.status(200).send({ token });
  } catch (e) {
    console.error(e);
    return res.status(401).send({ message: 'User not found.', e });
  }
};

export const authRequired = async (req, res, next) => {
  if (!req.headers.authorization) return res.status(401).end();

  const token = req.headers.authorization.split('Bearer ')[1];

  if (!token) {
    return res.status(401).end();
  }

  try {
    const payload = await verifyToken(token);
    const user = await User.findById(payload.id)
      .select('-password')
      .lean()
      .exec();

    req.user = user;
    next();
  } catch (e) {
    console.error(e);
    return res.status(401).end();
  }
};
