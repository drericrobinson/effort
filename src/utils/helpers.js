import { startOfWeek, endOfWeek, format } from 'date-fns';

export const weekStart = date => {
  return startOfWeek(date, { options: { weekStartsOn: 1 } });
};

export const weekEnd = date => {
  return endOfWeek(date, { options: { weekStartsOn: 1 } });
};

export const dateInWords = date => {
  return format(date, 'MMM D');
};

export const sendErrorMessage = (error, res) => {
  error.message === 'Unauthorized'
    ? res.status(401).json({
        error: "You don't have permission to access."
      })
    : res.status(400).end();
};
