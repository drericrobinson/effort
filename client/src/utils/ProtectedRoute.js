import React from 'react';
import { Redirect } from '@reach/router';
import { AuthConsumer } from './AuthContext';

export const ProtectedRoute = ({
  component: Component,
  altComponent: AltComponent = null,
  children,
  ...rest
}) => (
  <AuthConsumer>
    {({ isAuth }) =>
      isAuth ? (
        <Component {...rest}>{children}</Component>
      ) : AltComponent ? (
        <AltComponent {...rest}>{children}</AltComponent>
      ) : (
        <Redirect from='' to='login' noThrow />
      )
    }
  </AuthConsumer>
);
