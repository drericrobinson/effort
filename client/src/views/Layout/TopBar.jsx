import React from 'react';
import { Link } from '@reach/router';
import {
  UikTopBar,
  UikTopBarSection,
  UikTopBarTitle,
  UikTopBarLinkContainer,
  UikTopBarLink,
  UikDivider,
  UikButton
} from '../../@uik';
import { AuthConsumer } from '../../utils/AuthContext';
import { TopBarCoach } from './TopBarCoach';

export const TopBar = () => {
  return (
    <AuthConsumer>
      {({ logout, isCoach }) => (
        <UikTopBar>
          <UikTopBarSection>
            <Link to='/' style={{ textDecoration: 'none' }}>
              <UikTopBarTitle>Measure Effort</UikTopBarTitle>
            </Link>
            <UikDivider margin vertical />
            {isCoach() ? (
              <TopBarCoach />
            ) : (
              <UikTopBarLinkContainer>
                <UikTopBarLink
                  href='/report/current'
                  style={{ textDecoration: 'none' }}
                >
                  Current Report
                </UikTopBarLink>
                <UikTopBarLink
                  Component={Link}
                  to='/group'
                  style={{ textDecoration: 'none' }}
                >
                  Group
                </UikTopBarLink>
              </UikTopBarLinkContainer>
            )}
          </UikTopBarSection>

          <UikTopBarSection>
            <UikButton primary onClick={logout}>
              Logout
            </UikButton>
          </UikTopBarSection>
        </UikTopBar>
      )}
    </AuthConsumer>
  );
};
