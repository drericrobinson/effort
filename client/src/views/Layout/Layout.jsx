import React from 'react';
import { TopBar } from './TopBar';

export const Layout = props => {
  return (
    <div>
      <TopBar />
      <div style={{ maxWidth: '1200px', margin: '0 auto' }}>
        {props.children}
      </div>
    </div>
  );
};
