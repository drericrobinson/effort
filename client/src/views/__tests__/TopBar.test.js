import React from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import { TopBar } from '../Layout/TopBar';
import { AuthContext } from '../../utils/AuthContext';
import 'jest-dom/extend-expect';

afterEach(cleanup);

const mockLogout = jest.fn();

describe('<TopBar />', () => {
  it('should render correctly', () => {
    const tree = (
      <AuthContext.Provider value={{ logout: mockLogout }}>
        <TopBar />
      </AuthContext.Provider>
    );

    expect(render(tree)).toMatchSnapshot();
  });

  it('should be connected to AuthContext logout', () => {
    const tree = (
      <AuthContext.Provider value={{ logout: mockLogout }}>
        <TopBar />
      </AuthContext.Provider>
    );

    const { getByText } = render(tree);
    expect(getByText(/Logout/)).toBeInTheDocument();

    fireEvent.click(getByText(/Logout/));
    expect(mockLogout).toHaveBeenCalledTimes(1);
  });
});
