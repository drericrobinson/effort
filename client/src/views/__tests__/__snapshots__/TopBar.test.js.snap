// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopBar /> should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="uik-top-bar__wrapper"
      >
        <div
          class="uik-top-bar-section__wrapper"
        >
          <a
            aria-current="page"
            href="/"
            style="text-decoration: none;"
          >
            <h2
              class="uik-top-bar-title__wrapper"
            >
              Measure Effort
            </h2>
          </a>
          <div
            class="uik-divider__vertical uik-divider__margin"
          />
          <div
            class="uik-top-bar-link-container__wrapper"
          >
            <a
              class="uik-top-bar-link__wrapper"
              href="/report/current"
              style="text-decoration: none;"
            >
              Current Report
            </a>
            <a
              class="uik-top-bar-link__wrapper"
              href="/group"
              style="text-decoration: none;"
            >
              Group
            </a>
          </div>
        </div>
        <div
          class="uik-top-bar-section__wrapper"
        >
          <button
            class="uik-btn__base uik-btn__primary"
            type="button"
          >
            <span
              class="uik-btn__content"
            >
              Logout
            </span>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="uik-top-bar__wrapper"
    >
      <div
        class="uik-top-bar-section__wrapper"
      >
        <a
          aria-current="page"
          href="/"
          style="text-decoration: none;"
        >
          <h2
            class="uik-top-bar-title__wrapper"
          >
            Measure Effort
          </h2>
        </a>
        <div
          class="uik-divider__vertical uik-divider__margin"
        />
        <div
          class="uik-top-bar-link-container__wrapper"
        >
          <a
            class="uik-top-bar-link__wrapper"
            href="/report/current"
            style="text-decoration: none;"
          >
            Current Report
          </a>
          <a
            class="uik-top-bar-link__wrapper"
            href="/group"
            style="text-decoration: none;"
          >
            Group
          </a>
        </div>
      </div>
      <div
        class="uik-top-bar-section__wrapper"
      >
        <button
          class="uik-btn__base uik-btn__primary"
          type="button"
        >
          <span
            class="uik-btn__content"
          >
            Logout
          </span>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
